examples/pull.rs

```diff
@@ -0,0 +1,200 @@
+    println!("Fetching {} for repo", remote.name().unwrap());
+    remote.fetch(refs, Some(&mut fo), None)?;
+
+    {
+        // If there are local objects (we got a thin pack), then tell the user
+        // how many objects we saved from having to cross the network.
+        let stats = remote.stats();
+        if stats.local_objects() > 0 {
+            println!(
+                "\rReceived {}/{} objects in {} bytes (used {} local \
+                 objects)",
+                stats.indexed_objects(),
+                stats.total_objects(),
+                stats.received_bytes(),
+                stats.local_objects()
+            );
+        } else {
+            println!(
+                "\rReceived {}/{} objects in {} bytes",
+                stats.indexed_objects(),
+                stats.total_objects(),
+                stats.received_bytes()
+            );
+        }
+    }
+
+    let fetch_head = repo.find_reference("FETCH_HEAD")?;
+    Ok(repo.reference_to_annotated_commit(&fetch_head)?)
+}
+
+fn fast_forward(lb: &mut git2::Reference, rc: &git2::AnnotatedCommit) -> Result<(), git2::Error> {
+    let name = match lb.name() {
+        Some(s) => s.to_string(),
+        None => String::from_utf8_lossy(lb.name_bytes()).to_string(),
+    };
+    let msg = format!("Fast-Forward: Setting {} to id: {}", name, rc.id());
+    lb.set_target(rc.id(), &msg)?;
+    Ok(())
+}
+
+fn normal_merge(
+    repo: &Repository,
+    local: &git2::AnnotatedCommit,
+    remote: &git2::AnnotatedCommit,
+) -> Result<(), git2::Error> {
+    let local_tree = repo.find_tree(local.id())?;
+    let remote_tree = repo.find_tree(remote.id())?;
+    let ancestor = repo.find_tree(repo.merge_base(local.id(), remote.id())?)?;
+    let mut idx = repo.merge_trees(&ancestor, &local_tree, &remote_tree, None)?;
+
+    if idx.has_conflicts() {
+        println!("Merge conficts detected...");
+    }
+    let result_tree = repo.find_tree(idx.write_tree_to(repo)?)?;
+    // now create the merge commit
+    let msg = format!("Merge: {} into {}", remote.id(), local.id());
+    let sig = repo.signature()?;
+    let local_commit = repo.find_commit(local.id())?;
+    let remote_commit = repo.find_commit(remote.id())?;
+    // Do our merge commit and set current branch head to that commit.
+    let _merge_commit = repo.commit(
+        Some("HEAD"),
+        &sig,
+        &sig,
+        &msg,
+        &result_tree,
+        &[&local_commit, &remote_commit],
+    )?;
+    // Set working tree to match head.
+    repo.checkout_head(None)?;
+    Ok(())
+}
+
+fn do_merge<'a>(
+    repo: &'a Repository,
+    remote_branch: &str,
+    fetch_commit: git2::AnnotatedCommit<'a>,
+) -> Result<(), git2::Error> {
+    // 1. do a merge analysis
+    let analysis = repo.merge_analysis(&[&fetch_commit])?;
+
+    // 2. Do the appopriate merge
+    if analysis.0.is_fast_forward() {
+        // do a fast forward
+        match repo.find_reference(remote_branch) {
```

Comment:
I'm having a hard time following how this should work.  In my testing, the fast-forward never works because this line fails.  Calling `find_reference` on the string `master` isn't going to work.

From my limited understanding of how `pull` should work, shouldn't the fast-forward code path do the following?
1. Get the `HEAD` ref
2. Look up `FETCH_HEAD` (or maybe not necessary, since this has `fetch_commit`).
3. repo.checkout_tree(FETCH_HEAD)
4. Move HEAD to oid of FETCH_HEAD

Reply:
It is probably wrong :-D. I was very much feeling my way through the dark on these.

Reply:
Okay I've reloaded my state on this and I think my poor naming made it harder to reason through.

I could be wrong but I think the way this is supposed to work is that we want to set the reference we are pulling specifically the branch in your case 'master' to the result of the fetched commit. I should probably also be setting head as well which I don't think I'm doing but that isn't necessarily clear to me if that should happen or not.

I've updated the naming in this code to better indicate what I'm doing for clarity.